### ISSUE TITLE
Usage of Enums instead of flag

### ISSUE THREAD

**Issue by bzah** [cf-convention/discuss#238]
I could not find in the current CF convention or in the discussions here if netCDF4's Enums are forbidden or not. I know that CF currently recommend to use flag_meanings|flag_values|flag_mask (see [here ](https://github.com/cf-convention/cf-convention.github.io/blob/main/faq.md#how-can-i-encode-flag-values-or-other-enumerated-lists-with-cf) or [there](http://cfconventions.org/cf-conventions/cf-conventions.html#flags)), but there is a dedicated mean to define this kind of values in netCDF4 through Enum.

In short: I would like to know: **is it CF valid** to use netCDF4's **Enum** instead of a combination of _flag_meanings_ and _flag_values_ to describe an enumerated type ?

For reference:
- [Enums in netcdf C](https://docs.unidata.ucar.edu/netcdf-c/current/group__user__types.html#ga05a639c94ac2babcbeeb49eaae223322) 
- [Enums in netcdf Python](https://unidata.github.io/netcdf4-python/#enum-data-type)
- [Enums in netCDF Java](https://docs.unidata.ucar.edu/netcdf-java/current/userguide/cdm_datatypes.html#enumeration)

*Edit* After some archeology (thanks Lars!), there are existing issues about this topic:
- https://github.com/cf-convention/cf-conventions/issues/192
- https://github.com/cf-convention/cf-conventions/issues/191

**Comment by samain-eum** [comment]
> I think the answer is that it's _not_ CF-compliant to use netCDF4 Enums at the moment.

The question is a bit more subtle than that and I will provide context below. It is clear that currently enumerations are not part of the convention. So we cannot expect any reading tool relying on CF to do anything with the enumeration data type information (raw data decoding is handled by the base library though, so it just appears like a normal variable).

An enumerated variable may have additional flag_values and flag_meanings attributes, in which case it seems it would be CF compliant. It's an inelegant solution though as it would result in duplicating the information, with the risk of unconsistencies.

The use case we are having now at EUMETSAT, is that we have netCDF4 products using a lot of enumerations as this was deemed useful regardless of the CF-compliance aspect. Considering that changing formats at this stage would be massively inconvenient, we are trying to identify the best route to make them more CF-friendly in a pragmatic way. One way is to add the missing attributes (which won't break our existing interfaces). Another one is to adapt the tools we make available to our users (e.g. python readers) to be able to parse the enumeration datatype and create the same metadata as if the CF attributes were present. The latter case is more of a short term workaround though and the first one might be needed for compatibility with other tools.

**Comment by bzah** [comment]
Could this be a topic for the cf-workshop in October ?

**Comment by davidhassell** [comment]
I think this sounds like a good topic for the workshop.

**Comment by bzah** [comment]
We discussed this issue during the last CF workshop.
Tagging people from the workshop @MTG-Formats, @JonathanGregory, @davidhassell, @czender, @DocOtak 
It seems we agreed that Enums can be added to the convention.
This would require to:
- Review the whole convention and figure out what would be the impact of Enums for each chapter.
- Add a section to describe Enums themselves.
- Add a link to this section and explanations in flags chapter to describe how Enums can be an alternative to flag_meanings and flag_values.

Additionally we would like to see the support for Enums improved in various tools (The following is merely a wish list, not requirements):
- In netCDF C, we would like to be able to read variables that are normally typed with an Enum without actually reading the Enums and instead reading the underlying dtype.
The will be helpful for libraries that do not yet support Enums, so that they can read these variables.
- In netCDF C, we would like to see a proper workaround the issue of _FillValue with Enums (see https://github.com/Unidata/netcdf-c/issues/982#issuecomment-1708774123)
At the moment, when creating a varaible without setting the _FillValue, the variable may be filled with the default _FillValue of the underlying dtype (e.g. 255 for a ubyte).
This is an issue for Enums because this default _FillValue may not be a valid Enum value and will therefore crash even a standard tool like ncdump.
- Support the implementation or the raising of feature request for Enums in commonly used tools.


Notes:
- It would not be possible (or very difficult?) to localize Enums.
- Setting manually a FillValue for Enums may be performance intensive.
- Yet, we currently recommend to:
  - Have a value prepared for _FillValue in the Enum definition (like `"Missing": 0`)
  - Set the fill_value explicitly when creating the variable.

Please feel free to comment, I might have missed/forgotten something.

**Comment by ChrisBarker-NOAA** [comment]
> because the possible values and their meanings, for each Enum type, are defined only once in the file, not for each data variable, like for flag_values and flag_meanings.

I think it this is a more compelling argument for inclusion than either the data checking or ease for writers. It would could be helpful for consumers of the data as well to have it defined in one place. (that is, of there are use-cases -- I've never seen the same flag values used in multiple variables, but that's sut my limited experience)

> In netCDF C, we would like to be able to read variables that are normally typed with an Enum without actually reading the Enums and instead reading the underlying dtype. 

I'm confused -- can't you do that now? if it's stored as an int with flag-value, it can be read and cast to an Enum, yes?

At least I'm pretty sure that's what we do in some of our code now.

**Comment by JonathanGregory** [comment]
Dear Chris

> > In netCDF C, we would like to be able to read variables that are normally typed with an Enum without actually reading the Enums and instead reading the underlying dtype.
> 
> I'm confused -- can't you do that now? if it's stored as an int with flag-value, it can be read and cast to an Enum, yes?
> 
> At least I'm pretty sure that's what we do in some of our code now.

I have no experience of this myself, but our discussion group understood that the API would not let you obtain an Enum netCDF variable with the same function as you use for an integer netCDF variable, even though that is actually how it's stored on disk. That's my understanding of what was said, anyway. Abel @bzah or others more experienced can correct me!

Best wishes

Jonathan

**Comment by ChrisBarker-NOAA** [comment]
Ahh -- got it. I'm getting out of my area of expertise, but I *think* you can do essentially anything with C, but it would require custom code not built in to the netcdf C lib -- that makes sense. That could be a nice utility, yes.

**Comment by JonathanGregory** [comment]
Ideally, we were thinking, _existing_ code would require _no_ change to read Enums. If I understand correctly, that implies a change in the API. If that were done, and if Enum variables also had `flag_values` and `flag_meanings`, it would mean that _existing_ code would be able to read and interpret Enum variables without even being aware that they were Enums.

**Comment by czender** [comment]
If a client code that know nothing about the netCDF4 enum type enounters a file that stores an enum variable, then it will probably fail if it tries to read the enum. The reason is that the client would first need to use the API to figure out what the base type (integer, unsigned int, byte, etc) of the enum is so that the client can allocate enough memory before getting the variable. In language like C, all this needs to be done explicitly through the API. A smart interpreted language (e.g., Python) wrapper would do all this memory querying and allocation automatically but would probably still fail unless it had been specially coded to retrieve and treat the enum as the underlying base type rather than as a netcdf4 enum. My guess is that the functionality @JonathanGregory is wondering about would require new API functions e.g., `nc_get_var_as_base_type_rather_than_user_defined_type()`, and existing code would have to be recoded to always use those API functions rather than the default `nc_get_var()` functions.

**Comment by JonathanGregory** [comment]
Dear Charlie @czender 

Thanks for your explanation. That's helpful. Yes, I was hoping (in my comment here, and in the discussion) that _no_ change would be required to existing code. I suppose that means changing the effect of the existing get functions so that an Enum is fetched as its storage type. Then a different API enquiry would be needed to discover if this basic type is being used as an Enum.

However, if only a minimal change to existing code is required, such as you suggest, that would not be too bad as a price for supporting Enums in a minimal way.

Best wishes

Jonathan

**Comment by czender** [comment]
@JonathanGregory After further thought, I need to partially walk back my opinion (above) from yesterday. If a client program is "cautious enough" to allocate enough memory to hold the base type values of a netCDF4 enum variable, then it might succeed without any code modifications. Yesterday I said the program *must* know the base type of the enum. However, if the client just blindly reads the enum the standard nc_get_var() API and provides enough memory then it might work assuming that the netCDF library does not do a safety check that matches the type of the on-disk variable against the type of the buffer that the variable will be read into (the API requires specifying the type of destination buffer and it might use this information for safety checks). Whether netCDF does such a check, I don't know. Someone would need to test this to be sure. The library could, under the hood, just prevent the API from reading enums directly into integer types. Of course the attempt will definitely be buggy and likely fail if the client does not provide enough memory to hold the base type of the enum, e.g., if the client expects unsigned int32 and the base type is actually unsigned int64. Anyway, the situation is not quite as pessimistic as I thought yesterday.

**Comment by PythonCHB** [comment]
> Of course the attempt will definitely be buggy and likely fail if the client does not provide enough memory to hold the base type of the enum, e.g., if the client expects unsigned int32 and the base type is actually unsigned int64. 

This strikes me as simply the nature of C :-)

However, it seems the question, to some extent, is to what degree the netcdf-C lib should support CF out of the box -- I have no idea what Unidata thinks of this. Is there a C CF lib out there? 

Of course, Charlie has done a lot in NCO -- which is great!

**Comment by bzah** [comment]
I haven't been working on this topic as much as I should have, but from my understanding Enums do not yet meet the criteria to be added to CF. 

For this, we would need to have:
1. Good support from the netcdf reading libraries to read and write Enums.
2. Have mechanisms in these library to _seamlessly_ translate flag_values/flag_meanings into Enum and vice-versa.
3. Have more users, and not only one data-writer, who which to use Enums.

**[1]** and have been met in the python [xarray package](https://github.com/pydata/xarray/issues/8144), but this is merely one library and it required quite some work to get it merged. **[2]** was initially on the road plan for the same xarray PR, but we thought it would fit better on cf-xarray, so it's not yet implemented.
And, I suppose until **[3]** is reached, that is until more people are interested in using Enums, other libraries contributors will not spend the resources to implement such feature.

So, I suggest we close this issue for now, perhaps to revisit if Enums gain interest from other parties (or if EUMETSAT is willing to invest a lot in open-source contributions to make **[1]** and **[2]** possible @MTG-Formats)
